contracts/extensions/FeeTaker.sol

```diff
@@ -46,134 +52,108 @@ contract FeeTaker is IPostInteraction, IAmountGetter, Ownable {
-        }
+    function rescueFunds(IERC20 token, uint256 amount) external onlyOwner {
+        token.uniTransfer(payable(msg.sender), amount);
     }
 
     /**
      * @notice See {IPostInteraction-postInteraction}.
      * @dev Takes the fee in taking tokens and transfers the rest to the maker.
      * `extraData` consists of:
-     * 2 bytes — integrator fee percentage (in 1e5)
-     * 2 bytes — resolver fee percentage (in 1e5)
-     * 1 byte - taker whitelist size
-     * (bytes10)[N] — taker whitelist
+     * 1 byte - flags
      * 20 bytes — fee recipient
      * 20 bytes — receiver of taking tokens (optional, if not set, maker is used)
+     * 2 bytes — integrator fee percentage (in 1e5)
+     * 2 bytes — resolver fee percentage (in 1e5)
+     * bytes — whitelist structure determined by `_isWhitelistedPostInteractionImpl` implementation
+     * bytes — custom data to call extra postInteraction (optional)
      */
-    function postInteraction(
+    function _postInteraction(
         IOrderMixin.Order calldata order,
-        bytes calldata /* extension */,
-        bytes32 /* orderHash */,
+        bytes calldata extension,
+        bytes32 orderHash,
         address taker,
-        uint256 /* makingAmount */,
+        uint256 makingAmount,
         uint256 takingAmount,
-        uint256 /* remainingMakingAmount */,
+        uint256 remainingMakingAmount,
         bytes calldata extraData
-    ) external onlyLimitOrderProtocol {
+    ) internal virtual {
         unchecked {
-            (uint256 integratorFee, uint256 resolverFee, bytes calldata tail) = _parseFeeData(extraData, taker);
-            address feeRecipient = address(bytes20(tail));
-            tail = tail[20:];
+            bool customReceiver = extraData[0] & _CUSTOM_RECEIVER_FLAG == _CUSTOM_RECEIVER_FLAG;
+            address feeRecipient = address(bytes20(extraData[1:21]));
+            extraData = extraData[21:];
+
+            address receiver = order.maker.get();
+            if (customReceiver) {
+                receiver = address(bytes20(extraData));
+                extraData = extraData[20:];
+            }
+            (bool isWhitelisted, uint256 integratorFee, uint256 resolverFee, bytes calldata tail) = _parseFeeData(extraData, taker, _isWhitelistedPostInteractionImpl);
+            if (!isWhitelisted && _ACCESS_TOKEN.balanceOf(taker) == 0) revert OnlyWhitelistOrAccessToken();
+
             uint256 denominator = _FEE_BASE + integratorFee + resolverFee;
             // fee is calculated as a sum of separate fees to limit rounding errors
             uint256 fee = Math.mulDiv(takingAmount, integratorFee, denominator) + Math.mulDiv(takingAmount, resolverFee, denominator);
 
-            address receiver = order.maker.get();
-            if (tail.length > 0) {
-                receiver = address(bytes20(tail));
+            if (order.receiver.get() == address(this)) {
```

Comment:
Is it correct to use an order with the FeeTaker `postInteraction` extension and a different `receiver`? If they are different, all calculations in this method would be wasted, as the results are not propagated to further logic

Reply:
It is a workaround to allow orders with no fees order but all the other functionality enabled. I've added check for cases when fee > 0 but receiver is incorrectly set in 34234a3.